lndclient/lnd_services.go

```diff
@@ -183,6 +246,150 @@ func (s *GrpcLndServices) Close() {
 	log.Debugf("Lnd services finished")
 }
 
+// checkLndCompatibility makes sure the connected lnd instance is running on the
+// correct network, has the version RPC implemented, is the correct minimal
+// version and supports all required build tags/subservers.
+func checkLndCompatibility(conn *grpc.ClientConn, chainParams *chaincfg.Params,
+	readonlyMac serializedMacaroon, network string,
+	minVersion *verrpc.Version) (string, [33]byte, *verrpc.Version, error) {
+
+	// onErr is a closure that simplifies returning multiple values in the
+	// error case.
+	onErr := func(err error) (string, [33]byte, *verrpc.Version, error) {
+		closeErr := conn.Close()
+		if closeErr != nil {
+			log.Errorf("Error closing client connection: %v",
+				closeErr)
+		}
+
+		return "", [33]byte{}, nil, fmt.Errorf("error checking "+
+			"connected lnd version. at least version \"%s\" is "+
+			"required: %v", VersionString(minVersion), err)
+	}
+
+	// We use our own clients with a readonly macaroon here, because we know
+	// that's all we need for the checks.
+	lightningClient := newLightningClient(conn, chainParams, readonlyMac)
+	versionerClient := newVersionerClient(conn, readonlyMac)
+
+	// With our readonly macaroon obtained, we'll ensure that the network
+	// for lnd matches our expected network.
+	info, err := lightningClient.GetInfo(context.Background())
+	if err != nil {
+		return onErr(fmt.Errorf(
+			"unable to get info for lnd node: %v", err,
+		))
+	}
+	if network != info.Network {
+		return onErr(fmt.Errorf(
+			"network mismatch with connected lnd instance. "+
+				"wanted '%s', got '%s'", network, info.Network,
+		))
+	}
+
+	// Now let's also check the version of the connected lnd node.
+	err = checkVersionCompatibility(versionerClient, minVersion)
+	if err != nil {
+		return onErr(err)
+	}
+	version, err := versionerClient.GetVersion(context.Background())
+	if err != nil {
+		return onErr(fmt.Errorf(
+			"unable to get version for lnd node: %v", err,
+		))
+	}
+
+	// Return the static part of the info we just queried from the node so
+	// it can be cached for later use.
+	return info.Alias, info.IdentityPubkey, version, nil
+}
```

Comment:
While testing, I get the following message on lnd 0.9:
`error checking connected lnd version. at least version "v0.10.0, build tags 'signrpc,walletrpc,chainrpc,invoicesrpc'" is required: version check not implemented, need minimum lnd version of v0.10.0-beta`

not the most concise message for novice users

Reply:
Ok, I changed the error message a bit. It will now say: `lnd compatibility check failed: version check not implemented, need minimum lnd version of v0.10.0-beta`.

For the other errors (version mismatch or build tag missing), it will return the same, more detailed message: `error checking connected lnd version. at least version "v0.10.0, build tags 'signrpc,walletrpc,chainrpc,invoicesrpc'" is required`

Reply:
Tested, looks good